velox/exec/MergingVectorInput.cpp

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/exec/MergingVectorInput.h"
+#include "velox/exec/OperatorUtils.h"
+
+namespace facebook::velox::exec {
+
+MergingVectorInput::MergingVectorInput(
+    velox::memory::MemoryPool* pool,
+    int64_t preferredOutputBatchBytes,
+    int32_t preferredOutputBatchRows,
+    int32_t maxMergingRows,
+    bool preserveOrder)
+    : pool_(pool),
+      preferredOutputBatchBytes_(preferredOutputBatchBytes),
+      preferredOutputBatchRows_(preferredOutputBatchRows),
+      maxMergingRows_(maxMergingRows),
+      preserveOrder_(preserveOrder) {}
+
+RowVectorPtr MergingVectorInput::getVector(bool noMoreInput) {
+  if (noMoreInput) {
+    flush();
+  }
+  if (outputQueue_.size() > 0) {
+    const auto res = outputQueue_.front();
+    outputQueue_.pop();
+    return res;
+  }
+  return nullptr;
+}
+
+void MergingVectorInput::addVector(RowVectorPtr vector) {
+  if (nullptr == vector) {
```

Comment:
Can this be a VELOX_CHECK instead?

Reply:
If it used to merge output of FilterAndProject, the vector can be nullptr, so I add this check to ensure vector not nullptr.

Reply:
Let's just document that the caller needs to ensure that 'vector' is not null.

Reply:
Currently input=nullptr is normal, so just return directly.